### PR TITLE
Feature/room profile

### PR DESCRIPTION
**Pull request by ganfra**
This PR initialize the base for Room and RoomMember Profile screens. 
There is still actions to branch, but most of the P1 ones are ok. 

Screenshots: 

Room Profile: 
Screenshot_20200114-191706

Screenshot_20200114-191716

RoomMember Profile :
Screenshot_20200114-191738

### CODE REVIEW

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/DefaultProfileService.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                .configureWith(params) {
+                    this.callback = object : MatrixCallback<JsonDict> {
+                        override fun onSuccess(data: JsonDict) {
+                            val displayName = data[ProfileService.DISPLAY_NAME_KEY] as? String
+                            matrixCallback.onSuccess(Optional.from(displayName))
+                        }
+
+                        override fun onFailure(failure: Throwable) {
+                            matrixCallback.onFailure(failure)
+                        }
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/DefaultProfileService.kt]: Use `MatrixCallbackDelegate` to avoid this onFailure block

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/DefaultProfileService.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                .configureWith(params) {
+                    this.callback = object : MatrixCallback<JsonDict> {
+                        override fun onSuccess(data: JsonDict) {
+                            val avatarUrl = data[ProfileService.AVATAR_URL_KEY] as? String
+                            matrixCallback.onSuccess(Optional.from(avatarUrl))
+                        }
+
+                        override fun onFailure(failure: Throwable) {
+                            matrixCallback.onFailure(failure)
+                        }
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/DefaultProfileService.kt]: same remark

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/ProfileAPI.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+interface ProfileAPI {
+
+    /**
+     * Get the combined profile information for this user.
+     * This API may be used to fetch the user's own profile information or other users; either locally or on remote homeservers.
+     * This API may return keys which are not limited to displayname or avatar_url.
+     * @param userId the user id to fetch profile info
+     */
+    @GET(NetworkConstants.URI_API_PREFIX_PATH_R0 + "profile/{userId}")
+    fun getProfile(@Path("userId") roomId: String): Call<JsonDict>
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/ProfileAPI.kt]: roomId -> userId

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/ProfileModule.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        @Provides
+        @JvmStatic
+        @SessionScope
+        fun providesProfileAPI(retrofit: Retrofit): ProfileAPI {
+            return retrofit.create(ProfileAPI::class.java)
+        }
+    }
+
+    @Binds
+    abstract fun bindProfileService(userService: DefaultProfileService): ProfileService
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/ProfileModule.kt]: userService -> service

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/ProfileModule.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    @Binds
+    abstract fun bindGetProfileTask(getProfileInfoTask: DefaultGetProfileInfoTask): GetProfileInfoTask
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/ProfileModule.kt]: getProfileInfoTask -> task

File: vector/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -75,7 +75,7 @@
                 android:value=".features.home.HomeActivity" />
         </activity>
         <activity android:name=".features.debug.DebugMenuActivity" />
-        <activity android:name=".features.home.createdirect.CreateDirectRoomActivity" />
+        <activity android:name="im.vector.riotx.features.createdirect.CreateDirectRoomActivity" />
```
Comment (bmarty) [inline vector/src/main/AndroidManifest.xml]: Why this change?

File: vector/src/main/java/im/vector/riotx/core/epoxy/profiles/ProfileItemAction.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import com.airbnb.epoxy.EpoxyAttribute
+import com.airbnb.epoxy.EpoxyModelClass
+import im.vector.riotx.R
+import im.vector.riotx.core.epoxy.VectorEpoxyHolder
+import im.vector.riotx.core.epoxy.VectorEpoxyModel
+import im.vector.riotx.core.extensions.setTextOrHide
+import im.vector.riotx.features.themes.ThemeUtils
+
+@EpoxyModelClass(layout = R.layout.item_profile_action)
+abstract class ProfileItemAction : VectorEpoxyModel<ProfileItemAction.Holder>() {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/core/epoxy/profiles/ProfileItemAction.kt]: Naming convention, usually the name of the class is ended by Item. Rename to ProfileActionItem please

File: vector/src/main/java/im/vector/riotx/core/epoxy/profiles/ProfileItemSection.kt [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+import android.widget.TextView
+import com.airbnb.epoxy.EpoxyAttribute
+import com.airbnb.epoxy.EpoxyModelClass
+import im.vector.riotx.R
+import im.vector.riotx.core.epoxy.VectorEpoxyHolder
+import im.vector.riotx.core.epoxy.VectorEpoxyModel
+
+@EpoxyModelClass(layout = R.layout.item_profile_section)
+abstract class ProfileItemSection: VectorEpoxyModel<ProfileItemSection.Holder>() {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/core/epoxy/profiles/ProfileItemSection.kt]: Same remark about the naming

File: vector/src/main/java/im/vector/riotx/core/epoxy/profiles/ProfileMatrixItem.kt [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+@EpoxyModelClass(layout = R.layout.item_profile_matrix_item)
+abstract class ProfileMatrixItem : VectorEpoxyModel<ProfileMatrixItem.Holder>() {
+
+    @EpoxyAttribute lateinit var avatarRenderer: AvatarRenderer
+    @EpoxyAttribute lateinit var matrixItem: MatrixItem
+    @EpoxyAttribute var clickListener: View.OnClickListener? = null
+
+    override fun bind(holder: Holder) {
+        val bestName = matrixItem.getBestName()
+        val matrixId = matrixItem.id.takeIf { it != bestName }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/core/epoxy/profiles/ProfileMatrixItem.kt]: I love the way you use my MatrixItem 💋

File: vector/src/main/java/im/vector/riotx/features/home/room/list/actions/RoomListQuickActionsEpoxyController.kt [hunk 10]
```diff
@@ -55,11 +58,15 @@ class RoomListQuickActionsEpoxyController @Inject constructor(private val avatar
...
-        bottomSheetSeparatorItem {
-            id("leave_separator")
+        if (showAll) {
+            // Leave
+            dividerItem {
+                id("leave_separator")
+            }
+            RoomListQuickActionsSharedAction.Leave(roomSummary.roomId).toBottomSheetItem(5)
+
+            // Leave
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/features/home/room/list/actions/RoomListQuickActionsEpoxyController.kt]: remove this comment

File: vector/src/main/java/im/vector/riotx/features/navigation/DefaultNavigator.kt [hunk 11]
```diff
@@ -82,12 +84,9 @@ class DefaultNavigator @Inject constructor(
         }
     }
 
-    override fun openUserDetail(userId: String, context: Context, buildTask: Boolean) {
-        if (context is VectorBaseActivity) {
-            context.notImplemented("Open user detail")
-        } else {
-            context.toast(R.string.not_implemented)
-        }
+    override fun openRoomMemberProfile(userId: String, roomId: String?, context: Context, buildTask: Boolean) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/features/navigation/DefaultNavigator.kt]: unused param buildTask?

File: vector/src/main/java/im/vector/riotx/features/roommemberprofile/RoomMemberProfileFragment.kt [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+@file:Suppress("DEPRECATION")
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/features/roommemberprofile/RoomMemberProfileFragment.kt]: remove?

File: vector/src/main/java/im/vector/riotx/features/roommemberprofile/RoomMemberProfileViewEvents.kt [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.riotx.features.roommemberprofile
+
+/**
+ * Transient events for RoomProfile
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/features/roommemberprofile/RoomMemberProfileViewEvents.kt]: RoomProfile -> RoomMemberProfile

File: vector/src/main/java/im/vector/riotx/features/roommemberprofile/RoomMemberProfileViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+            }
+
+            override fun onFailure(failure: Throwable) {
+                _viewEvents.post(RoomMemberProfileViewEvents.Failure(failure))
+            }
+        }
+        if (isIgnored) {
+            session.unIgnoreUserIds(listOf(state.userId), ignoreActionCallback)
+        } else {
+            session.ignoreUserIds(listOf(initialState.userId), ignoreActionCallback)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/features/roommemberprofile/RoomMemberProfileViewModel.kt]: why using `state` and `initialState`?

File: vector/src/main/java/im/vector/riotx/features/roomprofile/RoomProfileFragment.kt [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+@file:Suppress("DEPRECATION")
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/features/roomprofile/RoomProfileFragment.kt]: remove

File: vector/src/main/java/im/vector/riotx/features/roomprofile/RoomProfileViewState.kt [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+import com.airbnb.mvrx.Async
+import com.airbnb.mvrx.MvRxState
+import com.airbnb.mvrx.Uninitialized
+import im.vector.matrix.android.api.session.room.model.RoomSummary
+
+data class RoomProfileViewState(
+        val roomId: String,
+        val roomSummary: Async<RoomSummary> = Uninitialized,
+        val isEncrypted: Boolean = false
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/features/roomprofile/RoomProfileViewState.kt]: This info is also in the roomSummary, no?

File: vector/src/main/res/values/strings_riotX.xml [hunk 17]
```diff
@@ -5,4 +5,31 @@
...
     <string name="login_error_threepid_denied">Your email domain is not authorized to register on this server</string>
 
+    <string name="room_profile_not_encrypted_subtitle">Messages in this room are not end-to-end encrypted.</string>
+    <string name="room_profile_encrypted_subtitle">Messages in this room are end-to-end encrypted.</string>
+    <string name="room_profile_section_security">Security</string>
+    <string name="room_profile_section_security_learn_more">Learn more</string>
+    <string name="room_profile_section_more">More</string>
+    <string name="room_profile_section_more_settings">Room settings</string>
+    <string name="room_profile_section_more_notifications">Notifications</string>
+    <string name="room_profile_section_more_member_list">"%1$s people"</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings_riotX.xml]: Shouldn't it be a plurals?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/profile/ProfileService.kt [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    companion object Constants {
+        const val DISPLAY_NAME_KEY = "displayname"
+        const val AVATAR_URL_KEY = "avatar_url"
+    }
+
+    fun getDisplayName(userId: String, matrixCallback: MatrixCallback<Optional<String>>): Cancelable
+
+    fun getAvatarUrl(userId: String, matrixCallback: MatrixCallback<Optional<String>>): Cancelable
+
+    fun getProfile(userId: String, matrixCallback: MatrixCallback<JsonDict>): Cancelable
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/profile/ProfileService.kt]: We can't, it's a totally dynamic hashmap...

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/DefaultProfileService.kt (hunk 1 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/profile/DefaultProfileService.kt]: This is not the same type

File: vector/src/main/AndroidManifest.xml (hunk 6 above)
Comment (ganfra) [reply vector/src/main/AndroidManifest.xml]: I think we should split the home package as it's not a "feature". But as you want

Comment (bmarty) [reply vector/src/main/AndroidManifest.xml]: I was talking about the fact that `im.vector.riotx` has been added. But ok for the move to another package.

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/profile/ProfileService.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/profile/ProfileService.kt]: ok
